### PR TITLE
Update analyzer to ^3.0.0-pre.20

### PR DESCRIPTION
**Pull request by bicknellr**
The new version of analyzer produces `Import`s where `document` may be undefined when it previously wouldn't have produced them at all, so this PR just throws in all of the places where `document` is used if `document` is undefined. Generally this is a good thing because it points out references to files that don't exist but it also means that workspace conversion is broken for the current set of repos listed in `docs/polymer-modulize-repos.txt` because some of these repos have incorrect references. I've been working on getting fixes into them but app-layout is one I'm not particularly sure what to do with: it contains a bunch of packages in [`/templates`](https://github.com/PolymerElements/app-layout/tree/0e237cdab14f2368f4f7fd3aa06af4ee5571c0ca/templates) which expect you to install things with bower and the files in those directories reach directly into `bower_components` to get them.

### CODE REVIEW

**Comment by bicknellr** [comment]
Updated so that at the places where imports are retrieved from analyzer, imports without a document are filtered out with a console warning.

File: src/document-converter.ts [hunk 1]
```diff
@@ -235,10 +248,18 @@ export class DocumentConverter extends DocumentProcessor {
     // itself.
     for (const scriptImport of this.document.getFeatures(
              {kind: 'html-script'})) {
+      if (scriptImport.document === undefined) {
+        console.warn(
+            `${this.originalPackageName} ${this.originalUrl}: ` +
+            `The script referenced using URL '${scriptImport.originalUrl}' ` +
+            `not be loaded and was ignored.`);
```
Comment (bicknellr) [inline src/document-converter.ts]: grammar: missing 'could'

File: src/package-scanner.ts [hunk 2]
```diff
@@ -270,7 +271,22 @@ export class PackageScanner {
...
+    const htmlImports = DocumentConverter.getAllHtmlImports(document);
+    const htmlImportsWithDocuments =
+        htmlImports.filter((htmlImport: Import) => {
+          const document = htmlImport.document;
+          const hasDocument = document !== undefined;
+          if (!hasDocument) {
+            console.warn(
+                `${packageName} ${documentUrl}: The document referenced ` +
+                `using URL '${
+                    htmlImport.originalUrl}' could not be loaded and ` +
```
Comment (bicknellr) [inline src/package-scanner.ts]: bad formatting

Comment (bicknellr) [reply src/package-scanner.ts]: fixed

File: src/document-converter.ts (hunk 1 above)
Comment (bicknellr) [reply src/document-converter.ts]: fixed

File: src/import-with-document.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import {Document, Import} from 'polymer-analyzer';
+
+export interface ImportWithDocument extends Import {
+  readonly document: Document;
+}
```
Comment (rictic) [inline src/import-with-document.ts]: Can also define a function that will tell you if you have an ImportWithDocument:

```
function importHasDocument(import_: Import): import_ is ImportWithDocument {
  return document !== undefined;
}
```

Then we can use this as a type safe alternative to casting with `as` (which should be avoided as much as possible).

Comment (bicknellr) [reply src/import-with-document.ts]: Very cool; I was trying out Flow recently and was trying to figure out how to write a type checking function but couldn't get past the fact that the narrowing only applied within the function. :/ It's neat that TypeScript has an explicit way to do this. I've updated the PR with an `isImportWithDocument` function.

**Comment by bicknellr** [comment]
Is it better to warn [like this](https://github.com/Polymer/polymer-modulizer/blob/bfcb8520fdc370914b22dd69707b3ff57f9e340a/src/document-processor.ts#L230)?

**Comment by bicknellr** [comment]
Tried it - yes it definitely is better.

**Comment by bicknellr** [comment]
I think this is ready for review again.

File: src/import-with-document.ts (hunk 3 above)
Comment (rictic) [reply src/import-with-document.ts]: That flow doc is neat!

I didn't realize that they did type checking of function parameters by looking at everywhere it's called. That seems like too much inference to me, but that might just be change aversion talking.

I really wish that `Array#filter` would flow through predicate functions in typescript.

File: src/document-converter.ts [hunk 4]
```diff
@@ -219,10 +219,14 @@ export class DocumentConverter extends DocumentProcessor {
     const filteredImports = [];
     for (const import_ of DocumentConverter.getAllHtmlImports(this.document)) {
       if (!isImportWithDocument(import_)) {
-        console.warn(
-            `${this.originalPackageName} ${this.originalUrl}: The document ` +
-            `referenced using URL '${import_.originalUrl}' could not be ` +
-            `loaded and was ignored.`);
+        console.warn(new Warning({
+          code: 'import-ignored',
+          message: `The import referenced by URL '${import_.originalUrl}' ` +
```
Comment (rictic) [inline src/document-converter.ts]: No need to include info that's present in the source code, it'll be logged with a squiggle under the original source code. So these messages could just be:

```
message: `Import could not be loaded and will be ignored.`,
```

File: src/import-with-document.ts (hunk 3 above)
Comment (bicknellr) [reply src/import-with-document.ts]: > I really wish that `Array#filter` would flow through predicate functions in typescript.

Same; this is actually why I changed `DocumentConverter#getHtmlImports` to a for loop in this PR.

File: src/document-converter.ts (hunk 4 above)
Comment (bicknellr) [reply src/document-converter.ts]: Ok, I'll update these messages and merge.

**Pull request merged**